src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+		// in this method.)
+		return method.AttributeLists.Any();
+	}
+
+	private static MethodDeclarationSyntax? ExtractSyntax(
+		GeneratorSyntaxContext context,
+		CancellationToken cancellationToken
+	) {
+		var methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+		var attributes = methodDeclaration.AttributeLists
+			.SelectMany( al => al.Attributes );
+
+		foreach( var attribute in attributes ) {
+			// We have the semantic model now, so make sure this method has
+			// the right attribute.
+			if( IsGenerateSyncAttribute( context.SemanticModel, attribute, cancellationToken ) ) {
+				return methodDeclaration;
+			}
+		}
+
+		return null;
+	}
+
+	internal static bool IsGenerateSyncAttribute(
+		SemanticModel model,
+		AttributeSyntax attribute,
+		CancellationToken cancellationToken
+	) {
+		var attributeConstructorSymbol = model.GetSymbolInfo( attribute, cancellationToken ).Symbol as IMethodSymbol;
+
+		if( attributeConstructorSymbol == null ) {
+			return false;
+		}
+
+		return attributeConstructorSymbol.ContainingType.ToDisplayString()
+			== "D2L.CodeStyle.Annotations.GenerateSyncAttribute";
+	}
+
+	private static MethodGenerationResult? GenerateSyncMethod(
+		(MethodDeclarationSyntax MethodDeclaration, Compilation Compilation) data,
+		CancellationToken cancellationToken
+	) {
+		if( !data.Compilation.ContainsSyntaxTree( data.MethodDeclaration.SyntaxTree ) ) {
+			return null;
+		}
+
+		var model = data.Compilation.GetSemanticModel( data.MethodDeclaration.SyntaxTree );
+		var methodSymbol = model.GetDeclaredSymbol( data.MethodDeclaration, cancellationToken );
+
+		if( methodSymbol == null || methodSymbol.Kind == SymbolKind.ErrorType ) {
+			return null;
+		}
+
+		var transformer = new AsyncToSyncMethodTransformer( model, cancellationToken );
+
+		var transformResult = transformer.Transform( data.MethodDeclaration );
+
+		return new(
+			Original: data.MethodDeclaration,
+			GeneratedSyntax: transformResult.Success ? transformResult.Value!.ToFullString() : "",
+			Diagnostics: transformResult.Diagnostics
+		);
+	}
+
+	private static IEnumerable<FileGenerationResult> GenerateFiles(
+		ImmutableArray<MethodGenerationResult> methodResults,
+		CancellationToken cancellationToken
+	) {
+		var methodsByFile = methodResults
+			.GroupBy( static mr => mr.Original.SyntaxTree.GetCompilationUnitRoot() )
+			.OrderBy( static mr => mr.First().Original.SyntaxTree.FilePath )
+			.ToImmutableArray();
+
+		var hintNames = SourceFilePathHintNameGenerator.Create(
+			// We're assuming every root has a tree here -- when is that false?
+			// Generated code?
+			methodsByFile.Select( static grp => grp.Key.SyntaxTree.FilePath ).ToImmutableArray()
+		).GetHintNames();
+
+		var files = methodsByFile.Zip(
+			hintNames,
+			static ( fileData, name ) => ( HintName: name, Data: fileData )
+		);
+
+		foreach( var file in files ) {
+			if( cancellationToken.IsCancellationRequested ) {
+				yield break;
+			}

```

Comment:
I believe it was determined msbuild expects and handles this?

```suggestion
			cancellationToken.ThrowIfCancellationRequested();
```

Reply:
It does, but I don't need to throw here.

Reply:
Under the assumption that roslyn will stop processing the enumerable produced by the SelectMany? Doesn't that same logic mean we don't need to check this at all, given we're a generator?

Reply:
Err, nvm for the second part that, I suppose generators aren't a state machine in that way in .NET. Had my implementations mixed up.

Reply:
`yield break` always stops enumeration. Generators are state machines, [here's a decompiled example that shows this](https://sharplab.io/#v2:D4AQDABCCMAsDcBYAUCkBmCBDARgZwBcAnLAYwKgCYIBhCAbxQmak10JPIhwHseAbCAGUAFjwCu/ACZCCPAA4AKCFIk5+AUwgAPCAEokyFhCYsMUaOgA8q8eo0A+CAFlFehqePNb9iAEsIAF4IMEMvFgB3ET9NZWJxLXdGI3CvGGgoAHZ/MNTjAIBqYPQAOjhcvOY/ADNlUQlpWQVlAPckz0rjdO4iDSwAawrKgF8O41GU5gnhoA)

It could be the case that the thing processing our resulting `IEnumerable<T>` (which is being transformed into an `IncrementalValuesProvider<T>`) also checks for cancellation, but there is "no way" for me to know that for sure.

It could be the case that their code doesn't check for cancellation at all, and that further parts of the pipeline continue to go on whatever comes out of our `IEnumerable<T>`, but that would be Microsoft's bug I believe.

The exception would be handy in places where we can't easily stop, e.g. inside `AsyncToSyncMethodTransformer` its going to be infeasible to stop the recursion, but in general control flow via exceptions kinda stinks.

Reply:
When you do absolutely want to throw is when you want to communicate in the standard way that the task was cancelled to the caller, but that's not really our responsibility here as far as I know, it's the frameworks.

Reply:
Sorry, I didn't communicate well since I was thinking through it in my head at the same time.

```csharp
		IncrementalValuesProvider<FileGenerationResult> generatedFiles =
			generatedMethods.Collect()
				.SelectMany( GenerateFiles );
```

What I was thinking through here is that while yes, `yield break` stops one invocation, it does not further invocations for other elements which would still do work.

However, SelectMany isn't Linq in this case, and it's actually a _1 => many (or none) transform_ (as per the comment in https://sourceroslyn.io/#Microsoft.CodeAnalysis/SourceGeneration/Nodes/ValueSourceExtensions.cs,5fdbcdb2b8308e69). Since it's a source of 1, there's no other invocations to make, so yes I agree it should be about equivalent.

Now, that was thinking I had to do though.

As far as I can tell, `ThrowIfCancellationRequested` is the idiom for cancellationTokens. `IsCancellationRequested` is there to check should you need to do cleanup / revert work before proceeding to ultimately call `ThrowIfCancellationRequested`.

See the example here, for instance: https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation

I would suggest just using `ThrowIfCancellationRequested` unless we've an explicit reason to do otherwise.

Reply:
We talked more on Zoom -- I'm still not 100% sold that this isn't allowed ala the first bullet point, and that it's the generators responsibility to not cache things mid cancellation, but it's at least not totally clear. I did notice that they check the tokens before they advance our iterator, but I don't want to trace around to see what happens _after_, and even so it's not clear thats a good idea (relying on implementation details? I'm not sure.)

Will switch to throw.